File: Application/khubdeemju/src/main/java/com/itsci/khubdeemju/model/RecordViolation.java
package com.itsci.khubdeemju.model;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.NoArgsConstructor;

import java.util.*;
import java.util.concurrent.ThreadLocalRandom;

@Data
@AllArgsConstructor
@NoArgsConstructor
@Entity
@Table(name = "RecordViolation")
@JsonIgnoreProperties({"hibernateLazyInitializer", "handler"})
public class RecordViolation {

    @Id
    @Column(name = "record_id", length = 10)
    private String record_id;
    @Column(name = "record_date", nullable = false)
    private Date record_date;
    @Column(name = "remaining_score", nullable = false)
    private int remaining_score;
    @Column(name = "picture_evidence")
    private String picture_evidence;
    @Column(name = "status_type", length = 50, nullable = false)
    private String status_type;
    @Column(name = "deduct_no", length = 2)
    private int deduct_no;
    @Column(name = "deduct_date")
    private Date deduc_date;

    @ManyToOne
    @JoinColumn(name = "student_id", nullable = false)
    private Student student;

    @ManyToOne
    @JoinColumn(name = "officer_id", nullable = false)
    private Officer officer;

    @ManyToOne
    @JoinColumn(name = "location_id", nullable = false)
    private Location location;

    @ManyToOne
    @JoinColumn(name = "violation_id", nullable = false)
    private ViolationType violationType;

//   สำหรับ Generate ID อัตโนมัติ
    @PrePersist
    public void generateRecordId() {
        if (this.record_id == null) {
            this.record_id = generateCustomId();
        }
    }

    private String generateCustomId() {
        int randomNum = ThreadLocalRandom.current().nextInt(1000, 9999); // สุ่มเลข 4 หลัก (0000 - 9999)
        return String.format("RC" + randomNum);
    }



}

File: Project_khubdee_mju_application/lib/model/recordviolation.dart
import 'package:project_final/model/location.dart';
import 'package:project_final/model/officer.dart';
import 'package:project_final/model/student.dart';
import 'package:project_final/model/violationtype.dart';

class RecordViolation {
  String? record_id;
  Student? student;
  DateTime? record_date;
  Officer? officer;
  Location? location;
  ViolationType? violationType;
  int? remaining_score;
  String? picture_evidence;
  String? status_type;
  int? deduct_no;
  DateTime? deduct_date;

  RecordViolation(
      {this.record_id,
      this.student,
      this.record_date,
      this.officer,
      this.location,
      this.violationType,
      this.remaining_score,
      this.picture_evidence,
      this.status_type,
      this.deduct_no,
      this.deduct_date});

  Map<String, dynamic> fromRecordViolationToJson() {
    return <String, dynamic>{
      'record_id': record_id,
      'student': student?.fromStudentToJson(),
      'record_date': record_date!.toIso8601String(),
      'officer': officer?.fromOfficerToJson(),
      'location': location?.fromLocationToJson(),
      'violationType': violationType?.fromViolationTypeToJson(),
      'remaining_score': remaining_score,
      'picture_evidence': picture_evidence,
      'status_type': status_type,
      'deduct_no': deduct_no,
      'deduct_date': deduct_date!.toIso8601String(),
    };
  }

  factory RecordViolation.fromJsonToRecordViolation(
          Map<String, dynamic> json) =>
      RecordViolation(
        record_id: json['record_id'],
        student: Student.fromJsonToStudent(json["student"]),
        record_date: DateTime.parse(json["record_date"]).toLocal(),
        officer: Officer.fromJsonToOfficer(json["officer"]),
        location: Location.fromJsonToLocation(json['location']),
        violationType:
            ViolationType.fromJsonToViolationType(json["violationType"]),
        remaining_score: json['remaining_score'],
        picture_evidence: json['picture_evidence'],
        status_type: json['status_type'],
        deduct_no: json['deduct_no'],
        deduct_date: json['deduct_date'] != null
            ? DateTime.parse(json["deduct_date"]).toLocal()
            : null,
      );
}
